components/eam/src/physics/cam/zm/zm_conv.F90

```diff
@@ -0,0 +1,2175 @@
+            s_int_g(i,k) = log(s_mid_g(i,k-1)/s_mid_g(i,k))*s_mid_g(i,k-1)*s_mid_g(i,k)/(s_mid_g(i,k-1)-s_mid_g(i,k))
+         else
+            s_int_g(i,k) = 0.5_r8*(s_mid_g(i,k)+s_mid_g(i,k-1))
+         end if
+         if (qdifr > interp_diff_min) then
+            q_int_g(i,k) = log(q_mid_g(i,k-1)/q_mid_g(i,k))*q_mid_g(i,k-1)*q_mid_g(i,k)/(q_mid_g(i,k-1)-q_mid_g(i,k))
+         else
+            q_int_g(i,k) = 0.5_r8*(q_mid_g(i,k)+q_mid_g(i,k-1))
+         end if
+      end do
+   end do
+
+   !----------------------------------------------------------------------------
+   ! calculate updraft and downdraft properties
+   call zm_cloud_properties(pcols, lengath, pver, pverp, msg, zm_param%limcnv, &
+                            p_mid_g, z_mid_g, z_int_g, t_mid_g, s_mid_g, s_int_g, q_mid_g, &
+                            landfrac_g, tpert_g, &
+                            msemax_klev_g, lel_g, jt, jlcl, j0, jd, &
+                            mflx_up, entr_up, detr_up, mflx_dn, entr_dn, mflx_net, &
+                            s_upd, q_upd, ql_g, s_dnd, q_dnd,  &
+                            q_mid_sat_g, cu_g, evp_g, pflx_g, rprd_g, &
+                            aero, loc_microp_st )
+
+   !---------------------------------------------------------------------------
+   ! convert from units of "per length" [1/m] to "per pressure" [1/mb].
+   do k = msg+1, pver
+      do i = 1,lengath
+         detr_up(i,k)= detr_up(i,k) * (z_int_g(i,k)-z_int_g(i,k+1))/p_del(i,k)
+         entr_up(i,k)= entr_up(i,k) * (z_int_g(i,k)-z_int_g(i,k+1))/p_del(i,k)
+         entr_dn(i,k)= entr_dn(i,k) * (z_int_g(i,k)-z_int_g(i,k+1))/p_del(i,k)
+         cu_g(i,k)   = cu_g(i,k)    * (z_int_g(i,k)-z_int_g(i,k+1))/p_del(i,k)
+         rprd_g(i,k) = rprd_g(i,k)  * (z_int_g(i,k)-z_int_g(i,k+1))/p_del(i,k)
+         evp_g(i,k)  = evp_g(i,k)   * (z_int_g(i,k)-z_int_g(i,k+1))/p_del(i,k)
+         if (zm_param%zm_microp) then
+            loc_microp_st%frz (i,k) = loc_microp_st%frz (i,k) * (z_int_g(i,k)-z_int_g(i,k+1))/p_del(i,k)
+            loc_microp_st%sprd(i,k) = loc_microp_st%sprd(i,k) * (z_int_g(i,k)-z_int_g(i,k+1))/p_del(i,k)
+         end if
+      end do
+   end do
+
+   !----------------------------------------------------------------------------
+   ! apply closure assumption to calculate cloud base mass flux
+   call zm_closure(pcols, lengath, pver, pverp, msg, cape_threshold_loc, &
+                   lcl_g, lel_g, jt, msemax_klev_g, dsubcld, &
+                   z_mid_g, z_int_g, p_mid_g, p_del, t_mid_g, &
+                   s_mid_g, q_mid_g, q_mid_sat_g, ql_g, s_int_g, q_int_g, &
+                   t_pcl_lcl_g, t_pcl_g, q_pcl_sat_g, s_upd, q_upd, &
+                   mflx_net, detr_up, mflx_up, mflx_dn, q_dnd, s_dnd, cape_g, &
+                   cld_base_mass_flux )
+
+   !----------------------------------------------------------------------------
+   ! limit cloud base mass flux to theoretical upper bound.
+   do i = 1,lengath
+      mflx_up_max(i) = 0
```

Comment:
For consistency, constant values 0 and 1 at lines 555, 560, and 574 should also add kind parameter _r8.

Reply:
fixed.